### DISCUSSION TITLE
Preconditioner and Linear solver

### DISCUSSION THREAD

**Discussion by AdelineHunter** [idaholab/moose#21601] [Q&A General] [answered]
Hi,

At first, I have some questions about the preconditioner and linear solver.

1. In some simple problems, we can use LU/iLU and `ksp gmres/bcgs` to solve linear equations. Unfortunately, LU could not solve complex problems due to expensive resource consumption, especially with more than 100000 DOFs. Therefore we must use other preconditioners such as `asm` or `bjacobi` to solve linear equations. Frustratingly, it often happens that the system of linear equations does not converge so that the whole problem cannot be solved. But I found that we can change some options such as `-sub_pc_factor_levels` and `-pc_factor_shift_type` to help solve linear equations. Are there any descriptions of these options like `-sub_pc_factor_levels` on any website?

2. I am always confused by diverged problems such as `DIVERGED_FNORM_NAN` and `DIVERGED_LINEAR_SEARCH`, etc.  Are there any descriptions of these diverged problems like `DIVERGED_FNORM_NAN` and `DIVERGED_LINEAR_SEARCH`, etc on any website?

Besides, I have some questions about the nonlinear solver.
1. MOOSE's default nonlinear solve method is PJFNK. In my opinion, if the jacobian matrix is could be made by hand or automatic difference, the NEWTON method is more effective and fast to solve the FEM problems. I'm not sure I'm right about the opinion.

2. In nonlinear solve progress, Is `linear_search` necessary? Are there any descriptions of these `line_search` methods on any website?

Thanks.

**Comment by GiudGiud** [comment, marked as answer]
Hello
1) 
See petsc documentation for more info. All we have on solver options is here:
https://mooseframework.inl.gov/source/executioners/Steady.html
2)
These are errors codes from Petsc that we are passing to the user. You may look at their documentation. They generally mean the nonlinear solver is failing, which either means there's a bug in the Jacobian or better initialization would help.
You should check the Jacobian with the Jacobian analyzer
https://mooseframework.inl.gov/help/development/analyze_jacobian.html

Nonlinear solver:
1) 
See this documentation
https://mooseframework.inl.gov/moose/source/systems/NonlinearSystem.html
Since we often use the Jacobian for the preconditioning in PJFNK, the memory consumption is similar to Newton's. But if we did not, the memory consumption would be much smaller for PJFNK than Newton's, hence the default. 

2)
All we have on line searches are this page. The petsc manual is a better reference since we're just importing a petsc tool here
https://mooseframework.inl.gov/syntax/LineSearches/
I turn off the line search for ALL my modeling. I've never had it improve things. But this is physics and numerics dependent. 

Guillaume

**Reply by AdelineHunter** [reply]
Thank you, Mr. Guillaume

Your advice is of great help to me.

According to my experience, the NEWTON method is the most efficient and effective method for solving nonlinear equations. PJFNK method is too slow. 

According to Krylov, I think there should be a way to tweak the JFNK method so that the residual drops even more, possibly at the cost of slower computation
![image](https://user-images.githubusercontent.com/108536058/180397950-55222eb6-a1fd-4174-a597-123b51d1b1bb.png)

What I want to know is whether there is such an adjustable interface to tune JFNK methods.

Thanks

**Reply by GiudGiud** [reply]
This is also going to live on the petsc side I'm afraid. See their documentation / their forum
